# Email Spam Detection using NLP 
### Building a Machine Learning Model for Text Classification

In [58]:
import pandas as pd 
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score , classification_report
from sklearn.model_selection import train_test_split
import re 
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords


## Dataset
The dataset consists of text samples and their corresponding labels.  
It is used to train a supervised machine learning model for classification tasks.

In [61]:
file_test = r'C:\Users\paton\OneDrive\Desktop\dataset\SMS_test.csv'
file_train = r'C:\Users\paton\OneDrive\Desktop\dataset\SMS_train.csv'


### Load Dataset
The dataset import to a pandas DataFrame to analysis.

In [64]:
data_1 = pd.read_csv(file_train, encoding="latin1")

### Data Overview
Exploration of dataset including the number of rows, columns, data types and missing value.

In [67]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 957 entries, 0 to 956
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   S. No.        957 non-null    int64 
 1   Message_body  957 non-null    object
 2   Label         957 non-null    object
dtypes: int64(1), object(2)
memory usage: 22.6+ KB


In [69]:
data_1.shape

(957, 3)

In [71]:
data_1.head()

,S. No.,Message_body,Label
0,1,Rofl. Its true to its name,Non-Spam
1,2,The guy did some bitching but I acted like i'd...,Non-Spam
2,3,"Pity, * was in mood for that. So...any other s...",Non-Spam
3,4,Will ü b going to esplanade fr home?,Non-Spam
4,5,This is the 2nd time we have tried 2 contact u...,Spam


In [73]:
data_1.isnull().sum()

S. No.          0
Message_body    0
Label           0
dtype: int64

## Data Preprocessing
Clean and prepare text data before modeling:
- Lowercasing
- Removing punctuation
- Removing stopwords
- Tokenization

In [76]:
data_1['Message_body'] = data_1['Message_body'].str.lower() 

In [78]:
nltk.download('stopwords')
en_stopword = stopwords.words('english')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\paton\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [80]:
data_1['Message_no_stopword'] = data_1['Message_body'].apply(lambda x:' '.join([word for word in x.split() if word not in en_stopword]))

In [82]:
data_1.head()

,S. No.,Message_body,Label,Message_no_stopword
0,1,rofl. its true to its name,Non-Spam,rofl. true name
1,2,the guy did some bitching but i acted like i'd...,Non-Spam,guy bitching acted like interested buying some...
2,3,"pity, * was in mood for that. so...any other s...",Non-Spam,"pity, * mood that. so...any suggestions?"
3,4,will ü b going to esplanade fr home?,Non-Spam,ü b going esplanade fr home?
4,5,this is the 2nd time we have tried 2 contact u...,Spam,2nd time tried 2 contact u. u £750 pound prize...


In [84]:
nltk.download('punkt_tab')
data_1['Message_Tokenize'] = data_1['Message_no_stopword'].apply(lambda x: word_tokenize(x))

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\paton\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [86]:
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()
data_1['Message_Cleaned'] = data_1['Message_Tokenize'].apply(
    lambda tokens: [lemmatizer.lemmatize(word) for word in tokens])

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\paton\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [88]:
data_1.head()

,S. No.,Message_body,Label,Message_no_stopword,Message_Tokenize,Message_Cleaned
0,1,rofl. its true to its name,Non-Spam,rofl. true name,"[rofl, ., true, name]","[rofl, ., true, name]"
1,2,the guy did some bitching but i acted like i'd...,Non-Spam,guy bitching acted like interested buying some...,"[guy, bitching, acted, like, interested, buyin...","[guy, bitching, acted, like, interested, buyin..."
2,3,"pity, * was in mood for that. so...any other s...",Non-Spam,"pity, * mood that. so...any suggestions?","[pity, ,, *, mood, that, ., so, ..., any, sugg...","[pity, ,, *, mood, that, ., so, ..., any, sugg..."
3,4,will ü b going to esplanade fr home?,Non-Spam,ü b going esplanade fr home?,"[ü, b, going, esplanade, fr, home, ?]","[ü, b, going, esplanade, fr, home, ?]"
4,5,this is the 2nd time we have tried 2 contact u...,Spam,2nd time tried 2 contact u. u £750 pound prize...,"[2nd, time, tried, 2, contact, u., u, £750, po...","[2nd, time, tried, 2, contact, u., u, £750, po..."


In [90]:
data_1['Message_Cleaned'] = data_1['Message_Cleaned'].apply(lambda x: ' '.join(x))

## Feature Engineering
Convert text data into numerical features using techniques such as:
- TF-IDF (Term Frequency - Inverse Document Frequency)

In [93]:
# Countvectorizing 
from sklearn.feature_extraction.text import TfidfVectorizer 

In [95]:
tfidfvec = TfidfVectorizer()
cout_vector_fit = tfidfvec.fit_transform(data_1['Message_Cleaned'])
print (cout_vector_fit)

  (0, 2189)	0.6629498232183731
  (0, 2675)	0.5348011938730758
  (0, 1759)	0.5239133658599108
  (1, 1211)	0.25979987514553854
  (1, 483)	0.36131085077803854
  (1, 274)	0.3408834861409456
  (1, 1529)	0.21569369000486108
  (1, 1381)	0.32639004027837465
  (1, 561)	0.32639004027837465
  (1, 2384)	0.27542550528672394
  (1, 936)	0.29146922977871087
  (1, 1789)	0.2632757415395664
  (1, 2827)	0.23018713586743192
  (1, 1140)	0.3151480502822953
  (1, 1097)	0.20914160298234016
  (2, 1961)	0.44890603025260817
  (2, 1715)	0.44890603025260817
  (2, 2570)	0.30798848210265567
  (2, 2372)	0.34816473822632577
  (2, 349)	0.44890603025260817
  (2, 2486)	0.4235263131806921
  (3, 1173)	0.3948726567837269
  (3, 964)	0.5750791031575305
  (3, 1094)	0.6095405863709299
  (3, 1293)	0.3765898081502929
  :	:
  (952, 2629)	0.20679978290434176
  (952, 1797)	0.2664317891277223
  (952, 1239)	0.30455316394672116
  (952, 1309)	0.2940633107041306
  (952, 2893)	0.3180769368846815
  (952, 1024)	0.3371376242941809
  (952, 177

In [97]:
tfidf_bag = pd.DataFrame(cout_vector_fit.toarray(), columns = tfidfvec.get_feature_names_out()) 
print (tfidf_bag)

     000  0121   02  0207  02073162414   03   04   05   06  07099833605  ...  \
0    0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
1    0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
2    0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
3    0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
4    0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
..   ...   ...  ...   ...          ...  ...  ...  ...  ...          ...  ...   
952  0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
953  0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
954  0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
955  0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   
956  0.0   0.0  0.0   0.0          0.0  0.0  0.0  0.0  0.0          0.0  ...   

     younger  yours  yoville   yr  yumm

## Model Building
Train machine learning models for text classification.  

In [100]:
model = GaussianNB()
y = data_1['Label']

In [102]:
X_train,X_test,y_train,y_test = train_test_split(tfidf_bag,y,test_size=0.3, random_state=7) 

In [104]:
model.fit(X_train,y_train)

GaussianNB()

In [106]:
y_pred = model.predict(X_test)
accuracy_score(y_pred,y_test)

0.9375

In [108]:
print (classification_report(y_pred,y_test))

              precision    recall  f1-score   support

    Non-Spam       0.96      0.97      0.96       247
        Spam       0.81      0.73      0.77        41

    accuracy                           0.94       288
   macro avg       0.88      0.85      0.87       288
weighted avg       0.94      0.94      0.94       288



## Model Evaluation

The model achieved an accuracy of **93.75%**.

- Non-Spam: Precision 0.96, Recall 0.97, F1-score 0.96  
- Spam: Precision 0.81, Recall 0.73, F1-score 0.77  

The model performs well overall, but has lower recall for spam, meaning some spam messages are misclassified.

In [111]:
import joblib

In [113]:
joblib.dump(model,'mail_classification.joblib')
joblib.dump(tfidfvec,'vector.joblib')

['vector.joblib']